src/tools/ToolSelectMove.ts

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * Enables Selection and Moving.
+ * Selections are additive with the shift key, and toggle with the option key.
+ * If a dot targeted on mouse down, that treats it as a selection, and transition to mouse move.
+ * If a dot is not targeted on mouse down, that starts a new selection.  Subclass this to create box or lasso selectors.
+ * If doing a mouse move, translate all the dots by the amount we've moved.
+ */
+export abstract class ToolSelectMove extends BaseMoveTool {
+  private moveToolStart: [number, number] | null = null;
+  protected selectionLassoStart: [number, number] | null = null;
+
+  onMouseDownInternal(event: MouseEvent): void {
+    const [x, y] = BaseTool.convertClientCoordinates(event);
+    const existingDotIndex = BaseTool.findDotAtEvent(event);
+
+    // reset what we are doing
+    this.moveToolStart = null;
+    this.selectionLassoStart = null;
+
+    if (existingDotIndex !== -1) {
+      // if we click on a selected dot, determine if we are toggling selection.
+      if (GlobalStore.state.selectedDots.includes(existingDotIndex)) {
+        if (event.altKey) {
+          GlobalStore.commit("toggleSelectedDots", [existingDotIndex]);
+        }
+      }
+      else {
+        if (!event.shiftKey) {
+          GlobalStore.commit("clearSelectedDots");
+        }
+        if (event.altKey) {
+          GlobalStore.commit("toggleSelectedDots", [existingDotIndex]);
+        } else {
+          GlobalStore.commit("addSelectedDots", [existingDotIndex]);
+        }
+      }
+      this.moveToolStart = [x, y];
+      this.doGrapherToolMoveAction([x, y]);
+    } else {
+      // if we hvae not clicked on a dot, start a new selection.
+      if (!event.shiftKey) {
+        GlobalStore.commit("clearSelectedDots");
+      }
+      GlobalStore.commit("setSelectionLasso", [[x, y]]);
+      this.selectionLassoStart = [x, y];
+    }
+  }
+
+  onMouseUpInternal(event: MouseEvent): void {
+    const [x, y] = BaseTool.convertClientCoordinates(event);
+
+    if (this.moveToolStart !== null) {
+      // We are in a Move, so translate all the dots.
+      const [deltaX, deltaY] = [
+        x - this.moveToolStart[0],
+        y - this.moveToolStart[1],
+      ];
+      const currentSSDots: StuntSheetDot[] =
+        GlobalStore.getters.getSelectedStuntSheet.stuntSheetDots;
+      GlobalStore.state.selectedDots.forEach((index: number) => {
+        const [roundedX, roundedY] = BaseTool.roundCoordinateToGrid([
+          currentSSDots[index].x + deltaX,
+          currentSSDots[index].y + deltaY,
+        ]);
+        GlobalStore.commit("moveDot", {
+          index: index,
+          position: [roundedX, roundedY],
+        });
+      });
+      // Set the ToolDots to be empty to indicate we're not moving anymore.
+      GlobalStore.commit("setGrapherToolDots", []);
+      // null out moveToolStart to incidcate we're not moving anymore.
+      this.moveToolStart = null;
+      return;
+    }
+    if (this.selectionLassoStart !== null) {
```

Comment:
nit: Could we turn this into an `else if` instead?
```suggestion
    } else if (this.selectionLassoStart !== null) {
```

Reply:
I really prefer early returns.  I like to have as much code as close to the left side of the page as possible. 

If you get to a point in a program you know that all the preconditions above have been resolved, so no reason to have an else statement.

So I'm going to leave as is.

Reply:
I prefer the style of early returns.  It allows code to be further on the left hand side, and makes things clearer that previous conditions have been resolved.